src/Analyzers/MSTest.Analyzers/DoNotUseShadowingAnalyzer.cs

```diff
@@ -0,0 +1,126 @@
+public sealed class DoNotUseShadowingAnalyzer : DiagnosticAnalyzer
+{
+    private static readonly LocalizableResourceString Title = new(nameof(Resources.DoNotUseShadowingTitle), Resources.ResourceManager, typeof(Resources));
+    private static readonly LocalizableResourceString Description = new(nameof(Resources.DoNotUseShadowingDescription), Resources.ResourceManager, typeof(Resources));
+    private static readonly LocalizableResourceString MessageFormat = new(nameof(Resources.DoNotUseShadowingMessageFormat), Resources.ResourceManager, typeof(Resources));
+
+    internal static readonly DiagnosticDescriptor DoNotUseShadowingRule = DiagnosticDescriptorHelper.Create(
+        DiagnosticIds.DoNotUseShadowingRuleId,
+        Title,
+        MessageFormat,
+        Description,
+        Category.Design,
+        DiagnosticSeverity.Warning,
+        isEnabledByDefault: true);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; }
+        = ImmutableArray.Create(DoNotUseShadowingRule);
+
+    public override void Initialize(AnalysisContext context)
+    {
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+        context.EnableConcurrentExecution();
+
+        context.RegisterCompilationStartAction(context =>
+        {
+            if (context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftVisualStudioTestToolsUnitTestingTestClassAttribute, out INamedTypeSymbol? testClassAttributeSymbol))
+            {
+                context.RegisterSymbolAction(
+                    context => AnalyzeSymbol(context, testClassAttributeSymbol),
+                    SymbolKind.NamedType);
+            }
+        });
+    }
+
+    private static void AnalyzeSymbol(SymbolAnalysisContext context, INamedTypeSymbol testClassAttributeSymbol)
+    {
+        var namedTypeSymbol = (INamedTypeSymbol)context.Symbol;
+        if (namedTypeSymbol.TypeKind != TypeKind.Class
+            || !namedTypeSymbol.GetAttributes().Any(attr => SymbolEqualityComparer.Default.Equals(attr.AttributeClass, testClassAttributeSymbol)))
+        {
+            return;
+        }
+
+        Dictionary<string, List<ISymbol>> symbolGroups = new();
+
+        INamedTypeSymbol? currentType = namedTypeSymbol.BaseType;
+        while (currentType is not null)
+        {
+            foreach (ISymbol member in currentType.GetMembers())
+            {
+                if (member.IsOverride || member.Name == ".ctor")
+                {
+                    continue;
+                }
+
+                if (!symbolGroups.TryGetValue(member.Name, out List<ISymbol>? members))
+                {
+                    members = new List<ISymbol>();
+                    symbolGroups[member.Name] = members;
+                }
+
+                // Add the member to the list
+                members.Add(member);
+            }
+
+            currentType = currentType.BaseType;
+        }
```

Comment:
Let's have a method for this part so that the main method is composing readable blocks.